src/lib/rule2E.ts

```diff
@@ -0,0 +1,187 @@
+import {Context, getDefaultContext} from './context';
+import {computeTextAlternative} from './compute_text_alternative';
+
+// Input types that imply role 'textbox' if list attribute is not present,
+// and imply role 'combobox' if list attribute is present.
+const TEXT_INPUT_TYPES = ['email', 'tel', 'text', 'url', 'search'];
+
+/**
+ * Determines whether a given node has role 'textbox' and,
+ * if so, gets the value of that textbox.
+ * @param node - element whose role is being calculated
+ * @return - textbox value if node is a textbox, null otherwise
+ * (null indicates that node is not a textbox).
+ */
+function getValueIfTextbox(node: HTMLElement): string | null {
+  // Explicit role='textbox' (elements not of type <input>) handled by rule2F
+
+  // type <textarea> implies role='textbox'
+  if (node instanceof HTMLTextAreaElement) {
+    return node.value;
+  }
+
+  if (!(node instanceof HTMLInputElement)) {
+    return null;
+  }
+
+  // <input> with certain type values & no list attribute implies role='textbox'
+  const nodeInputType = node.getAttribute('type')?.toLowerCase() ?? '';
+  if (TEXT_INPUT_TYPES.includes(nodeInputType) && !node.hasAttribute('list')) {
+    return node.value;
+  }
+
+  return null;
+}
+
+/**
+ * Determines whether a given node has role 'combobox'
+ * or 'listbox' and, if so, gets the text alternative for the
+ * option(s) selected by that combobox / listbox.
+ * @param node - node whose role is being calculated
+ * @param context - information relevant to the calculation of that role
+ * @return - text alternative for selected option(s) if node is a
+ * combobox or listbox, null otherwise.
+ * (null indicates that node is neither combobox nor listbox).
+ */
+function getValueIfComboboxOrListbox(
+  node: HTMLElement,
+  context: Context
+): string | null {
+  // Combobox role implied by input type and presence of list attribute,
+  // chosen option is the input value.
+  if (node instanceof HTMLInputElement) {
+    const nodeInputType = node.getAttribute('type')?.toLowerCase() ?? '';
+    if (TEXT_INPUT_TYPES.includes(nodeInputType) && node.hasAttribute('list')) {
+      return node.value;
+    }
+  }
+
+  // Listbox may be defined explicitly using 'role',
+  // and using 'aria-selected' attribute to mark selected options.
+  const nodeRole = node.getAttribute('role') ?? '';
+  if (nodeRole.toLowerCase() === 'listbox') {
+    // #SPEC_ASSUMPTION (E.2) : consider multiple selected options' text
+    // alternatives, joining them with a space as in 2B.ii.c
+    const selectedOptionsTextAlternative = Array.from(node.childNodes)
+      .map(childNode => {
+        if (
+          childNode instanceof HTMLElement &&
+          childNode.getAttribute('aria-selected') === 'true'
+        ) {
```

Comment:
you can do this directly using `node.querySelectorAll('[role="option"][aria-selected="true"]')`